### ISSUE TITLE
Jetty Http2 client timeouts

### ISSUE THREAD

**Issue by mobile-bit** [eclipse/jetty.project#2066]
Hello All,

We use Jetty Http2 client(9.4.3.v20170317) to send messages to an HTTP2 server(we dont have access to it). 

Some of the requests just timeout and we are not able to figure the reason out. Its not reproducible and this happens only in our production server. 

(I have scrubbed sensitive information from the logs. Let me know if any extra information is needed)

I have captured the Debug log of jetty below to show the failed request. After HTTP2Flusher Appends the HeadersFrame, nothing happens.

In case of successful request the HTTPFlusher Flushes the HTTP2ClientSession and continues
`org.eclipse.jetty.http2.HTTP2Flusher: Flushing HTTP2ClientSession@3023c6d4`

Can you please provide any hint/ pointers to understand what the issue could be ? 

Thanks

```
12 Dec 2017 20:46:59,762 [DEBUG]  org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /3/sign/6D2C0C1F05E1567F8EC5BC178B36F2006D2C0C1F05E1567F8EC5BC178B36F200 HTTP/1.1]@20a1717 for HttpDestination[https://my.server.com]@3398c22d,queue=1,pool=MultiplexConnectionPool@3697fa6a[c=10/10,b=0,m=1,i=9]
12 Dec 2017 20:46:59,763 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 org.eclipse.jetty.client.AbstractConnectionPool: Connection active HttpConnectionOverHTTP2@73b0e46c(closed=false)[HTTP2ClientSession@3023c6d4{l:/10.134.43.195:45725 <-> r:my.server.com/19.188.123.146:443,queueSize=300,sendWindow=0,recvWindow=63597,streams=300,NOT_CLOSED}]
12 Dec 2017 20:46:59,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpDestination: Processing exchange HttpExchange@2db92673 req=PENDING/null@null res=PENDING/null@null on HttpConnectionOverHTTP2@73b0e46c(closed=false)[HTTP2ClientSession@3023c6d4{l:/10.134.43.195:45725 <-> r:my.server.com/19.188.123.146:443,queueSize=320,sendWindow=0,recvWindow=63597,streams=300,NOT_CLOSED}] of HttpDestination[https://my.server.com]@3398c22d,queue=0,pool=MultiplexConnectionPool@3697fa6a[c=10/10,b=0,m=1,i=9]
12 Dec 2017 20:46:59,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpChannel: HttpExchange@2db92673 req=PENDING/null@null res=PENDING/null@null associated true to HttpChannelOverHTTP2@62724164(exchange=HttpExchange@2db92673 req=PENDING/null@null res=PENDING/null@null)
12 Dec 2017 20:46:59,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpSender: Request begin HttpRequest[POST /3/device/6D2C0C1F05E1567F8EC5BC178B36F2006D2C0C1F05E1567F8EC5BC178B36F200 HTTP/2.0]@20a1717
12 Dec 2017 20:46:59,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpSender: Request headers HttpRequest[POST /3/device/6D2C0C1F05E1567F8EC5BC178B36F2006D2C0C1F05E1567F8EC5BC178B36F200 HTTP/2.0]@20a1717
12 Dec 2017 20:46:59,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.io.IdleTimeout: HTTP2Stream@70257b49#690055{sendWindow=0,recvWindow=0,reset=false,NOT_CLOSED} idle timeout check, elapsed: 0 ms, remaining: 3600000 ms
12 Dec 2017 20:46:59,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.http2.HTTP2Session: Created local HTTP2Stream@70257b49#690055{sendWindow=65535,recvWindow=65535,reset=false,NOT_CLOSED}
12 Dec 2017 20:46:59,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.http2.HTTP2Flusher: Appended HeadersFrame@224568f5#690055{end=false}, frames=371
12 Dec 2017 20:47:09,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpExchange: Failed HttpExchange@2db92673 req=COMPLETED/java.util.concurrent.TimeoutException@462b4870 res=COMPLETED/java.util.concurrent.TimeoutException@462b4870: req=true/rsp=true java.util.concurrent.TimeoutException
12 Dec 2017 20:47:09,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpSender: Request failure HttpRequest[POST /3/device/6D2C0C1F05E1567F8EC5BC178B36F2006D2C0C1F05E1567F8EC5BC178B36F200 HTTP/2.0]@20a1717 HttpExchange@2db92673 req=COMPLETED/java.util.concurrent.TimeoutException@462b4870 res=COMPLETED/java.util.concurrent.TimeoutException@462b4870 on HttpChannelOverHTTP2@62724164(exchange=HttpExchange@2db92673 req=COMPLETED/java.util.concurrent.TimeoutException@462b4870 res=COMPLETED/java.util.concurrent.TimeoutException@462b4870): java.util.concurrent.TimeoutException
12 Dec 2017 20:47:09,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpExchange: Terminated request for HttpExchange@2db92673 req=TERMINATED/java.util.concurrent.TimeoutException@462b4870 res=COMPLETED/java.util.concurrent.TimeoutException@462b4870, result: null
12 Dec 2017 20:47:09,764 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpSender: Terminating request HttpRequest[POST /3/device/6D2C0C1F05E1567F8EC5BC178B36F2006D2C0C1F05E1567F8EC5BC178B36F200 HTTP/2.0]@20a1717
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpReceiver: Response failure HttpResponse[null 0 null]@36a258b0 HttpExchange@2db92673 req=TERMINATED/java.util.concurrent.TimeoutException@462b4870 res=COMPLETED/java.util.concurrent.TimeoutException@462b4870 on HttpChannelOverHTTP2@62724164(exchange=HttpExchange@2db92673 req=TERMINATED/java.util.concurrent.TimeoutException@462b4870 res=COMPLETED/java.util.concurrent.TimeoutException@462b4870): java.util.concurrent.TimeoutException
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpExchange: Terminated response for HttpExchange@2db92673 req=TERMINATED/java.util.concurrent.TimeoutException@462b4870 res=TERMINATED/java.util.concurrent.TimeoutException@462b4870, result: Result[HttpRequest[POST /3/device/6D2C0C1F05E1567F8EC5BC178B36F2006D2C0C1F05E1567F8EC5BC178B36F200 HTTP/2.0]@20a1717 > HttpResponse[null 0 null]@36a258b0] java.util.concurrent.TimeoutException
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpReceiver: Response complete HttpResponse[null 0 null]@36a258b0
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpChannel: HttpExchange@2db92673 req=TERMINATED/java.util.concurrent.TimeoutException@462b4870 res=TERMINATED/java.util.concurrent.TimeoutException@462b4870 disassociated true from HttpChannelOverHTTP2@62724164(exchange=null)
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpDestination: Released HttpConnectionOverHTTP2@73b0e46c(closed=false)[HTTP2ClientSession@3023c6d4{l:/10.134.43.195:45725 <-> r:my.server.com/19.188.123.146:443,queueSize=26,sendWindow=0,recvWindow=63444,streams=297,NOT_CLOSED}]
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpReceiver: Request/Response failed: Result[HttpRequest[POST /3/device/6D2C0C1F05E1567F8EC5BC178B36F2006D2C0C1F05E1567F8EC5BC178B36F200 HTTP/2.0]@20a1717 > HttpResponse[null 0 null]@36a258b0] java.util.concurrent.TimeoutException
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.http2.HTTP2Session: Sending ResetFrame@6d06821e#690055{cancel_stream_error}
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.http2.HTTP2Flusher: Appended ResetFrame@6d06821e#690055{cancel_stream_error}, frames=27
12 Dec 2017 20:47:09,765 [DEBUG] d232dee6-d913-4103-b4dd-184668802864 ( : 17) org.eclipse.jetty.client.HttpExchange: Aborted (true) while active HttpExchange@2db92673 req=TERMINATED/java.util.concurrent.TimeoutException@462b4870 res=TERMINATED/java.util.concurrent.TimeoutException@462b4870: java.util.concurrent.TimeoutException
```

**Comment by gregw** [comment]
While we first the log, any chance you can try the latest 9.4.8 release?

**Comment by mobile-bit** [comment]
Yes, I’m importing the latest 9.4.8 and update here once I have tested with it.

**Comment by sbordet** [comment]
@mobile-bit you have news ?

**Comment by mobile-bit** [comment]
Imported the latest release and it’s going through our pre-production stages. Should have it tested in the next couple of days.

**Comment by mobile-bit** [comment]
@sbordet no luck, even with the latest release, i still see timeouts.
Some observations during this event:
1. High volume, this is common with a spike in requests.
2. I found this pattern with few timeout request: Not sure if relevant
```
12 Dec 2017 20:46:59,764 org.eclipse.jetty.io.IdleTimeout: HTTP2Stream@70257b49#690055{sendWindow=0,recvWindow=0,reset=false,NOT_CLOSED} idle timeout check, elapsed: 0 ms
, remaining: 3600000 ms
.
.
.
12 Dec 2017 20:46:59,764 org.eclipse.jetty.http2.HTTP2Session: Created local HTTP2Stream@70257b49#690055{sendWindow=65535,recvWindow=65535,reset=false,NOT_CLOSED}
.
.
.
intermingled with other requests
.
.
.
12 Dec 2017 20:46:59,876 org.eclipse.jetty.http2.HTTP2Stream: Update close for HTTP2Stream@70257b49#690055{sendWindow=65535,recvWindow=65535,reset=false,NOT_CLOSED} close=false local=true
```

Please let me know what I can look for to get more information.

**Comment by sbordet** [comment]
@mobile-bit the "pattern" you describe above is normal.

In your first message, it is strange that the `HTTP2Flusher` appends a `HEADERS` frame, but then there appears to be no processing of that frame.

If you can, please log at DEBUG level category `org.eclipse.jetty`, so that we will have more information about non-HTTP2 components that may be involved in this issue.

What would also be of great help is to call `HTTP2Client.dump()` when you see that the system becomes unresponsive.

**Comment by mobile-bit** [comment]
@sbordet The first log i pasted was at debug level and there are too many org.eclipse.jetty logs from other requests. I am not able to identify the logs pertaining to the failed request. Can you mention any specific thing i can look for to pin it down to the failing requests like stream number, session,... ?

I did a HttpClient dump, we use HttpClientTransportOverHTTP2
```
 += com.xx.xx.xx.http2.PingSendingHttpClientTransportOverHTTP2@2316758f - STARTED
 |   += org.eclipse.jetty.http2.client.HTTP2Client@11090320 - STARTED
 |       +- com.xx.xx.xx.http2.Http2ClientFactory$1@66b98a51
 |       +~ QueuedThreadPool@HttpClient@1457780431{STARTED,8<=8<=200,i=6,q=0} - STARTED
 |       +~ org.eclipse.jetty.util.thread.ScheduledExecutorScheduler@6a7cf049 - STARTED
 |       += org.eclipse.jetty.http2.client.HTTP2Client$ClientSelectorManager@7be7d12e - STARTED
 |       |   += ReservedThreadExecutor@761498a3{s=1/8,p=0}@org.eclipse.jetty.http2.client.HTTP2Client$ClientSelectorManager@7be7d12e - STARTED
 |       |   += org.eclipse.jetty.io.ManagedSelector@663e21e3 id=0 keys=1 selected=0 actions=0 - STARTED
 |       |       += EatWhatYouKill@3f772e07/SelectorProducer@7a5cf666/PRODUCING/ReservedThreadExecutor@761498a3{s=1/8,p=0}@org.eclipse.jetty.http2.client.HTTP2Client$ClientSelectorManager@7be7d12e - STARTED
 |       |           +- SelectorProducer@7a5cf666
 |       |       +- keys
 |       |       |   +- sun.nio.ch.EPollSelectorImpl@126c695f keys=1
 |       |       |   +- SelectionKey@df7c50c{i=1}->SocketChannelEndPoint@1cab8562{my.server.com/19.188.123.146:443<->/10.156.194.139:34473,OPEN,fill=FI,flush=-,to=60/3600000}{io=1/1,kio=1,kro=1}->SslConnection@1894022f{NOT_HANDSHAKING,eio=-1/-1,di=-1}=>HTTP2ClientConnection@5d968f59
 |       |       +- actions
 |       +- org.eclipse.jetty.http2.client.http.HttpClientTransportOverHTTP2$$Lambda$49/1344209971@4478a80
 |       += ReservedThreadExecutor@285fb80e{s=0/0,p=0}@null - STARTED
 |       += HTTP2ClientSession@3f79b5df{l:/10.156.194.139:34473 <-> r:my.server.com/19.188.123.146:443,queueSize=0,sendWindow=65444,recvWindow=16777216,streams=0,NOT_CLOSED} - STARTED
 |           +- BufferingFlowControlStrategy@3978a321[ratio=0.50,sessionLevel=0,sessionStallTime=0ms,streamsStallTime=0ms]
 +- org.eclipse.jetty.util.SocketAddressResolver$Async@248a8dae
 += HttpDestination[https://my.server.com]@71c0def9,queue=0,pool=MultiplexConnectionPool@3e55a53[c=1/10,b=0,m=0,i=1] - STARTED
     +- MultiplexConnectionPool@3e55a53[c=1/10,b=0,m=0,i=1]
         +- HttpConnectionOverHTTP2@636c89b4(closed=false)[HTTP2ClientSession@3f79b5df{l:/10.156.194.139:34473 <-> r:my.server.com/19.188.123.146:443,queueSize=0,sendWindow=65444,recvWindow=16777216,streams=0,NOT_CLOSED}][0]
     +- exchanges
```

Does this provide any clue ?

**Comment by sbordet** [comment]
@mobile-bit don't worry about too many log lines; please grab the logs as detailed as possible and we will figure them out. Don't remove log lines, just grab them from a while before the issue happens, to after the issue happens. If it's big, zip it and attach it as a file here. Point being that we know how to look at the logs and what to search for, so we will do the job of figuring out things. Try to not modify/redact the logs; if you do, make sure that no line is deleted.

The `HttpClient.dump()` looks normal and shows a completely idle `HttpClient`: no exchanges queued, no HTTP/2 streams pending, no frames in the HTTP/2 session queue.

In your first comment, the logs seemed to show that a `HEADERS` frame was queued in the session, but then not processed. This would have shown up in the dump, but it does not, so the dump did not capture the issue.

**Comment by mobile-bit** [comment]
@sbordet I am still trying to get approval for posting the production logs. 
In the meantime, i observed one more kind of timeout:

I configured the maxConnectionPerRoute to be 10 and at times of spikes the HTTPDestination queue shoots to 600 and this would cause the connection acquisition to fail both times in `activate` and `trycreate`.

https://github.com/eclipse/jetty.project/blob/21365234f8cf036829f7a03e5c7b89a9572444f9/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java#L97

Increasing the maxConnectionPerRoute to 600 solves this kind of timeout. 

I was of the opinion that Multiplex connection will have upto 1000(maxMulitplex - in our server's case) streams and 10,000 requests could be served by this client, but as these requests are happening at the same time, every request is acquiring a new connection from the MutliplexConnectionPool. As the previous connection per route was limited to 10 , acquiring the connection was silently failing after 2 attempts. 
Is my understanding correct ?

https://github.com/eclipse/jetty.project/blob/21365234f8cf036829f7a03e5c7b89a9572444f9/jetty-client/src/main/java/org/eclipse/jetty/client/MultiplexConnectionPool.java#L142

**Comment by sbordet** [comment]
You don't want to have `maxConnectionPerDestination` at 600 with HTTP/2.
Your problem is of a different kind and increase that value only hides it.

If you have `maxConnectionsPerDestination=10` and `maxConcurrentStreams=1000` then you can handle 10000 _concurrent_ requests.
That means that you can have 10000 _outstanding_ requests that did not get a response yet.

Sending a second request does not open a second connection. Rather, the first connection is used and a new multiplexed request is sent over it.
Only when you reach `maxConcurrentStreams` a new connection is opened.

The client-side idle timeout for a request is really large at 1 hour.
If you're stuck with a lot of requests unresponded, then you have to double check your server - it may not complete the responses and the client remains hung.

**Comment by mobile-bit** [comment]
Thanks for the immediate response !

thats interesting, is there a way to prove that the delay is coming from the server ?  I can add debug statements to the jetty client.

Can you please explain this behavior:
On sending the Http Request, the request is queued as an exchange in https://github.com/eclipse/jetty.project/blob/21365234f8cf036829f7a03e5c7b89a9572444f9/jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java#L256

The queue size of the destination is only 770 according to the debug log when we fail to acquire the client but the MultiplexConnectionPool is `[c=10/10,b=0,m=0,i=0]` . 

1. _Shouldn't the queue size for this `HttpDestination` be 10*1000 for the connections to be exhausted_ ?
2. _Why are muxed, busy and idle all connections at 0_ ?

```
HttpDestination[https://my.server.com]@27814b4f,queue=770,pool=MultiplexConnectionPool@4ed7617b[c=10/10,b=0,m=0,i=0]
org.eclipse.jetty.client.MultiplexConnectionPool: <myDebugStatement> muxedConnections is empty
org.eclipse.jetty.client.AbstractConnectionPool: Max connections 10/10 reached
org.eclipse.jetty.client.MultiplexConnectionPool: <myDebugStatement> muxed is empty
```

**Comment by sbordet** [comment]
Requests are always queued, but only remain in the `HttpDestination` queue if they cannot be sent, meaning there are no connections available and no streams slots available in the connections.

You can add a timeout to the requests:

```java
httpClient.newRequest(...)
        .timeout(15, TimeUnit.SECONDS)
```

If you see the requests failing because of timeouts, it means the server does not reply in time.

**Comment by mobile-bit** [comment]
@sbordet do u see a reason for `MultiplexConnectionPool@4ed7617b[c=10/10,b=0,m=0,i=0]` all connections being 0 and the connections all being used ?

**Comment by sbordet** [comment]
@mobile-bit no, that sounds like a bad state. I would have expected `i=10`.

Can you provide the logs, at least those for the connection pool ?

**Comment by mobile-bit** [comment]
the  log looks like this:

```
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 199) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@76e21d3 for HttpDestination[https://my.server.com]@c3e0a99,queue=2,pool=MultiplexConnectionPool@ea014b6[c=0/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 199) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 199) org.eclipse.jetty.client.AbstractConnectionPool: Connection 1/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 721) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@747b5273 for HttpDestination[https://my.server.com]@c3e0a99,queue=2,pool=MultiplexConnectionPool@ea014b6[c=0/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 721) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 721) org.eclipse.jetty.client.AbstractConnectionPool: Connection 2/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 199) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 721) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 6bf2402cdf28 (Thread : 1056) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@4e1663bd for HttpDestination[https://my.server.com]@c3e0a99,queue=3,pool=MultiplexConnectionPool@ea014b6[c=2/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] 6bf2402cdf28 (Thread : 1056) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 6bf2402cdf28 (Thread : 1056) org.eclipse.jetty.client.AbstractConnectionPool: Connection 3/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] 6bf2402cdf28 (Thread : 1056) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 09be0373eb11 (Thread : 573) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@233f2dc9 for HttpDestination[https://my.server.com]@c3e0a99,queue=4,pool=MultiplexConnectionPool@ea014b6[c=3/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] 09be0373eb11 (Thread : 573) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 09be0373eb11 (Thread : 573) org.eclipse.jetty.client.AbstractConnectionPool: Connection 4/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] 09be0373eb11 (Thread : 573) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 119) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@22054eeb for HttpDestination[https://my.server.com]@c3e0a99,queue=5,pool=MultiplexConnectionPool@ea014b6[c=4/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 119) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 119) org.eclipse.jetty.client.AbstractConnectionPool: Connection 5/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] 1fa31fdeb31c (Thread : 119) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 1004) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@7a7aa2ee for HttpDestination[https://my.server.com]@c3e0a99,queue=7,pool=MultiplexConnectionPool@ea014b6[c=5/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] 4b83e3c1db70 (Thread : 200) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@538d00b1 for HttpDestination[https://my.server.com]@c3e0a99,queue=7,pool=MultiplexConnectionPool@ea014b6[c=5/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 1004) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 1004) org.eclipse.jetty.client.AbstractConnectionPool: Connection 6/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 1004) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 91aa822bdcc0 (Thread : 465) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@4a480d8f for HttpDestination[https://my.server.com]@c3e0a99,queue=8,pool=MultiplexConnectionPool@ea014b6[c=6/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] 91aa822bdcc0 (Thread : 465) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 91aa822bdcc0 (Thread : 465) org.eclipse.jetty.client.AbstractConnectionPool: Connection 7/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 591) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@2a748cc0 for HttpDestination[https://my.server.com]@c3e0a99,queue=10,pool=MultiplexConnectionPool@ea014b6[c=7/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 370) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@3120beff for HttpDestination[https://my.server.com]@c3e0a99,queue=10,pool=MultiplexConnectionPool@ea014b6[c=7/10,b=0,m=0,i=0]
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 591) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 591) org.eclipse.jetty.client.AbstractConnectionPool: Connection 8/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 370) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 370) org.eclipse.jetty.client.AbstractConnectionPool: Connection 9/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] 4b83e3c1db70 (Thread : 200) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 4b83e3c1db70 (Thread : 200) org.eclipse.jetty.client.AbstractConnectionPool: Connection 10/10 creation
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 591) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 4b83e3c1db70 (Thread : 200) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] db22ee9a2ecd (Thread : 370) org.eclipse.jetty.client.MultiplexConnectionPool: MuxConn Empty 2000
09 Jan 2018 05:08:52,617 [DEBUG] 0d8318747356 (Thread : 616) org.eclipse.jetty.client.HttpDestination: Queued HttpRequest[POST /5/msg/xxx HTTP/1.1]@18942f46 for HttpDestination[https://my.server.com]@c3e0a99,queue=12,pool=MultiplexConnectionPool@ea014b6[c=10/10,b=0,m=0,i=0]
```
[contextFile.txt](https://github.com/eclipse/jetty.project/files/1620803/contextFile.txt)

**Comment by sbordet** [comment]
The logs show `HttpClient` trying to open connections to `HttpDestination[https://my.server.com]@c3e0a99`, but it never succeeds in opening a connection, and the server does not reply to the request to open a connection.

Granted, the logs do not show enough time to tell whether the server eventually grants opening a connection, but the client will eventually timeout following `HttpClient.connectTimeout` (by default 15 seconds).

That also explains why you see `MultiplexConnectionPool@4ed7617b[c=10/10,b=0,m=0,i=0]`: none of the connections has actually been established yet.

You have to dig into why the server does not accept connections.

**Comment by mobile-bit** [comment]
I dont see even a single creation failed logs - https://github.com/eclipse/jetty.project/blob/21365234f8cf036829f7a03e5c7b89a9572444f9/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java#L122
Irrespective of acquiring the connection or not, requests timeout and also I observed that even though i set maxConnections to 600 , only the idle connections go up and muxedconnections remain 1. So the publish path does not seem to have any issue as the exchange gets into the queue. 

I have to prove that the server is taking time/failing to respond and want to make sure Jetty is not causing this issue.  
I will next look into 
1. logs around reception path from selector to HttpReceiver https://github.com/eclipse/jetty.project/blob/de7e3001efcc970a5e4bd63371f8ea93b6893e2c/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java#L406 
2. Want to make sure the thread count is sufficient for our spikes - https://github.com/eclipse/jetty.project/blob/21365234f8cf036829f7a03e5c7b89a9572444f9/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java#L95

_Can you recommend any way to see from the networking perspective or Jetty logs, that the requests were sent successfully but the server is failing to respond_

Thanks

**Comment by sbordet** [comment]
What I meant to say is that the server is taking a long time to accept connections at the TCP level.

That is why you don't see connect failures on the client (you have to wait the connect timeout to see those).

If you have a large connect timeout, and a shorter timeout for requests, the requests may fail while still queued in the destination.

I would not concentrate much on the client, the problem is on the server or on the network between the two.

Try to put a short connect timeout (e.g. 2 seconds) on the client and see if the client reports connect failures.
If it does, you have to understand why it takes more than 2 seconds to connect to the server.
Could be firewalls, transparent proxies, load balancers, overloaded server, etc.

**Comment by mobile-bit** [comment]
Sorry, I am still not able to post the logs, thanks a lot for all the responses. 

The connect time is generally well within 1 second. I tried increasing the total time for every request from 10 seconds to 120 seconds, just to see if the requests eventually completes and that did it, all the requests completed. Though some of the requests took more than 70 seconds to finish. I am pretty convinced that the publish path looks good, it is only with the reception.

Now, I need to make sure the delay is not coming from Jetty or our JVM configuration or our network settings before I can push the issue on to Server. 

1. _Do you think increasing the `QueuedThreadPool` from 200 will improve the reading rate (if it is bottle neck)_ ?
2. Can you recommend any way to make sure that our JVM or network settings are not the reason for congestion ? 

The reason I want to double check our network settings is because this happens only when we have a spike in requests and not during other times.

**Comment by sbordet** [comment]
> I am pretty convinced that the publish path looks good, it is only with the reception.

The logs you provided so far show that is (also) the sending part, and in particular opening TCP connection, that takes time - and this is due to the server.

I would not change the number of threads unless you have identified what is the problem in your system.
I doubt very much that the problem is on the client.

**Comment by mobile-bit** [comment]
What is an acceptable time (avg or p99 ) for creating a new connection ? Sometimes i see the connections are created within 1 millisecond other times its around 2 seconds.

**Comment by sbordet** [comment]
Connection creation time is only the TCP time: `SYN + SYN/ACK + ACK`.

**Comment by nezihyigitbasi** [comment]
@mobile-bit were you able to get to the bottom of this issue?

**Comment by mobile-bit** [comment]
@nezihyigitbasi Sorry for the late response..

After a long back & forth between Networking team and server team, I did 2 things:

1. Added time logging for different request(Request ) callbacks 
`request.onRequestHeaders(new Request.HeadersListener()`
`request.onComplete(new Response.CompleteListener() `
`request.onRequestQueued(new Request.QueuedListener()`
`request.onResponseBegin(new Response.BeginListener()`

I saw high latency between requestHeaders and responseBegin. Jumped to conclusion that it was server that is taking time. After some intense debugging from server side they were convinced that it is either network or our application.

2. At this point started doing a TcpDump of the http2 tls traffic and using wire shark started analyzing the Timeout packages. To my surprise, found that _After the header is sent, the data packet is not being sent for 23 seconds. Once the data packet is sent the server responds within 300 ms_. 

Questions:
1. Any reason why this might be happening from jetty perspective ? 
2. Can onRequestContent​(Request.ContentListener listener) be used to find this behavior ? 


Thanks for responding to my question. I am hoping to get to the bottom of the issue.

**Comment by sbordet** [comment]
@mobile-bit what you found seems a different problem, as what reported by @nezihyigitbasi clearly showed a TCP problem at accepting connections, confirmed by the logs.

@mobile-bit when you say "after the header is sent" you mean the HTTP/2 `HEADERS` frame, and by "data packet" you mean the HTTP/2 `DATA` frame ?

If that's the case, that may be due to the fact that the HTTP/2 flow control window is zero, and only frees up after 23 seconds. You should be able to see this in JMX or in a dump() by looking at the session's stalled times.
If you don't see flow control stall times, then it's another problem like thread exhaustion or something similar.

Listener `Request.ContentListener` can be used to track when `DATA` frames are sent.
You should see a 23 s delay between `Request.HeadersListener` and `Request.ContentListener`.

Do you see retransmissions in the TCP dump ?

**Comment by mobile-bit** [comment]
@sbordet Yes HTTP/2 HEADERS with 
* no `END_STREAM` 
* yes `END_HEADERS`

and `HTTP/2 DATA` after 20 seconds with 
* yes `END_STREAM`

I see very few retransmissions(like 30 over 2 hours) in wireshark with `tcp.analysis.retransmission` filter. 

I will check JMX and flow control window using a dump and update here.

Thanks

**Comment by mobile-bit** [comment]
@sbordet 
5 mins before the `DATA` frame being blocked I see a `WINDOW_UPDATE ` frame with Window Size Increment of 16711681 is sent. Looks like the default is 65535 

Does this mean there is a flow control stall ?

**Comment by mobile-bit** [comment]
@sbordet I did a client dump when timeouts happen and found the following line in there,

`BufferingFlowControlStrategy@306d9e06[ratio=0.50,sessionLevel=0,sessionStallTime=15428ms,streamsStallTime=0ms]
`
and one more with `BufferingFlowControlStrategy@157af2d7[ratio=0.50,sessionLevel=0,sessionStallTime=25826ms,streamsStallTime=0ms]`

Does this mean we are having Flow Control issue ? if so how do i work around it ? 

Thanks

**Comment by mobile-bit** [comment]
One more update:

We are sending 400 requests in a given second and each request is approximately 200 bytes. 
The first 96 requests succeed and rest of them timeout. The first 96 request mostly took 300 ms to complete. There are more than 300 requests in the next 300 ms making the send window to go over 65535 Bytes. 

Even though there could be a max of 1000 requests but the sendWindow needs to be within 65535 bytes. So, how do we deal with this scenario ?

I am going to increase `HttpClient.requestBufferSize` and see if that improves the situation. Can you recommend any other way to deal with this issue ?

Thanks

**Comment by sbordet** [comment]
@mobile-bit the default flow control window size in the HTTP/2 specification is 65535 bytes, but in Jetty (both client and server) it has been enlarged to MiB.
The client has a default session flow control window of 16 MiB, so 400 requests at 200 bytes of request content gives 80k which is well within 16 MiB.

`HttpClient.requestBufferSize` is not used in HTTP/2.

Seeing session stall times at 15 s and 25 s is not good, it means that (cumulatively) that session was slowed down considerably because of flow control.
Since `streamsStallTime=0`, it means that the streams never saw flow control stall, so this is a case of a session having many streams that collectively exhaust the session flow control window.

Before changing the `BufferingFlowControlStrategy` I would double check the session flow control window configuration, since the numbers do not figure.

**Comment by mobile-bit** [comment]
@sbordet 
What kind of setting should be checked for session flow control ? 


We dont have any special setting:
We use the client in the following way:
1. pass `HTTP2Client` to `HttpClientTransportOverHTTP2` 
2. use the transport in `HttpClient` 

We had a override for ByteBufferPool for both `HTTP2Client` and `HttpClient`. Not sure why it was put in place. I have removed it and going to try today. 

```
public ByteBuffer acquire(int capacity, boolean direct) {
            ByteBuffer byteBuffer = direct ? ByteBuffer.allocateDirect(capacity) : ByteBuffer.allocate(capacity);
            byteBuffer.limit(0);
            return byteBuffer;
        }
```

Thanks

**Comment by sbordet** [comment]
@mobile-bit the default should cover most cases for flow control settings.
I agree you should be just using the default buffer pools.

**Comment by mobile-bit** [comment]
@sbordet It looks like removing the overridden ByteBufferPool is improving the timeouts. 
I will do more testing to make sure we see less timeouts with this change.

Is there any reason why this could have caused increased timeouts (Session stall time) ?

Thanks

**Comment by sbordet** [comment]
@mobile-bit cannot think of any reason why the `acquire()` method you posted above would increase the timeouts.

**Comment by mobile-bit** [comment]
@sbordet the byte buffer was a false alarm, it did not have any impact. 
Coming back to session stall time, is there any way to improve that ? 

```
 I would double check the session flow control window configuration, since the numbers do not figure.
```
Anything specific ? Everything is default.
Can we create new connections in `MultiplexConnectionPool` based on the flow control rather than the number of requests(stream count) ?

**Comment by sbordet** [comment]
@mobile-bit sorry we need to consolidate this. It started with 9.4.3 with a timeout, but the dump was normal, then apparently a TCP connectivity issue, then it was a flow control window configuration, then buffers, etc.

The flow control windows should be at 16 MiB for the session and 8 MiB for the stream so reading 64 KiB from your logs means that something is wrong in the configuration.

Jetty 9.4.9 is just out with a lot of improvements in logging and dump, so can you please try that out ?
We need a detailed scenario where you have the failure, possibly with logs and dump.

**Comment by sbordet** [comment]
Lack of activity, closing. Please drop a message if you have more information and you want to reopen this issue.

**Comment by ranjithpp** [comment]
I am experiencing similar issues

2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Connection: HTTP2 onFillable HTTP2ClientConnection@2b4243<-SocketChannelEndPoint@11f9258{/10.53.133.90:16800<->/10.43.18.48:49325,OPEN,fill=-,flush=-,to=0/30000}{io=1/0,kio=1,kro=1}->HTTP2ClientConnection@2b4243 
2019-02-08 12:32:46,461 DEBUG o.e.j.u.t.ReservedThreadExecutor: ReservedThreadExecutor@12ecab3{s=1/2,p=0}@1faf279 task=EatWhatYouKill@13c2de7/SelectorProducer@1feb2d5/IDLE/p=true/QueuedThreadPool[qtp2193062]@2176a6{STARTED,8<=8<=200,i=5,q=0}[pc=0,pic=1,pec=0,epc=18]@2019-02-08T12:32:46.461+05:30
2019-02-08 12:32:46,461 DEBUG o.e.j.u.t.s.EatWhatYouKill: EatWhatYouKill@b2becc/HTTP2Producer@1b13feb/IDLE/p=false/QueuedThreadPool[qtp2193062]@2176a6{STARTED,8<=8<=200,i=5,q=0}[pc=0,pic=0,pec=0,epc=0]@2019-02-08T12:32:46.461+05:30 tryProduce false
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Connection: Dequeued task null
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ChannelEndPoint: filled 77 SocketChannelEndPoint@11f9258{/10.53.133.90:16800<->/10.43.18.48:49325,OPEN,fill=-,flush=-,to=0/30000}{io=1/0,kio=1,kro=1}->HTTP2ClientConnection@2b4243
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Connection: Filled 77 bytes
2019-02-08 12:32:46,461 DEBUG o.e.j.h.p.Parser: Parsed HEADERS frame header from java.nio.HeapByteBuffer[pos=9 lim=77 cap=8192]
2019-02-08 12:32:46,461 DEBUG o.e.j.h.h.HpackDecoder: CtxTbl[848fc4] decoding 7 octets
2019-02-08 12:32:46,461 DEBUG o.e.j.h.h.HpackDecoder: decode 88Be0f0d023532
2019-02-08 12:32:46,461 DEBUG o.e.j.h.h.HpackDecoder: decode IdxStatic {S,8,:status: 200(evaluated),4e9d87}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.h.HpackDecoder: decode Be0f0d023532
2019-02-08 12:32:46,461 DEBUG o.e.j.u.t.s.EatWhatYouKill: EatWhatYouKill@13c2de7/SelectorProducer@1feb2d5/IDLE/p=true/QueuedThreadPool[qtp2193062]@2176a6{STARTED,8<=8<=200,i=5,q=0}[pc=0,pic=1,pec=0,epc=18]@2019-02-08T12:32:46.461+05:30 tryProduce true
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: updateable 0
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: updates 0
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ChannelEndPoint: Key interests updated 1 -> 0 on SocketChannelEndPoint@11f9258{/10.53.133.90:16800<->/10.43.18.48:49325,OPEN,fill=-,flush=-,to=0/30000}{io=0/0,kio=0,kro=1}->HTTP2ClientConnection@2b4243
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: Selector sun.nio.ch.WindowsSelectorImpl@2ff3f8 waiting on select
2019-02-08 12:32:46,461 DEBUG o.e.j.h.h.HpackDecoder: decode Idx {D,0,content-type: text/html; charset=UTF-8,434ec}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.h.HpackDecoder: decode 0f0d023532
2019-02-08 12:32:46,461 DEBUG o.e.j.h.h.HpackDecoder: decoded 'content-length: 52' by IdxName/LitVal/
2019-02-08 12:32:46,461 DEBUG o.e.j.h.p.BodyParser: Parsed HEADERS frame hpack from java.nio.HeapByteBuffer[pos=16 lim=77 cap=8192]
2019-02-08 12:32:46,461 DEBUG o.e.j.h.c.HTTP2ClientSession: Received HeadersFrame@f36dc0#35{end=false}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Stream: Update close for HTTP2Stream@ca613#35{sendWindow=65535,recvWindow=65535,reset=false,LOCALLY_CLOSED,age=101,attachment=null} close=false local=false
2019-02-08 12:32:46,461 INFO c.b.j.h.s.HTTP2StreamHandler: TEST: onHeaders: 200frame.isEndStream() false
2019-02-08 12:32:46,461 DEBUG o.e.j.h.p.Parser: Parsed HEADERS frame body from java.nio.HeapByteBuffer[pos=16 lim=77 cap=8192]
2019-02-08 12:32:46,461 DEBUG o.e.j.h.p.Parser: Parsed DATA frame header from java.nio.HeapByteBuffer[pos=25 lim=77 cap=8192]
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Session: Received DataFrame@9616e0#35{length:52,end=true}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.FlowControlStrategy: Data received, 52 bytes, updated session recv window 16776332 -> 16776280 for HTTP2ClientSession@1e7f928{l:/10.43.18.48:49325 <-> r:/10.53.133.90:16800,sendWindow=65535,recvWindow=16776280,streams=3,NOT_CLOSED,null}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.FlowControlStrategy: Data received, 52 bytes, updated stream recv window 65535 -> 65483 for HTTP2Stream@ca613#35{sendWindow=65535,recvWindow=65483,reset=false,LOCALLY_CLOSED,age=101,attachment=null}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Stream: Update close for HTTP2Stream@ca613#35{sendWindow=65535,recvWindow=65483,reset=false,LOCALLY_CLOSED,age=102,attachment=null} close=true local=false
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Session: Removed local HTTP2Stream@ca613#35{sendWindow=65535,recvWindow=65483,reset=false,CLOSED,age=102,attachment=null}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.FlowControlStrategy: Data consumed, 52 bytes, session recv window level 936/8388608 for HTTP2ClientSession@1e7f928{l:/10.43.18.48:49325 <-> r:/10.53.133.90:16800,sendWindow=65535,recvWindow=16776280,streams=2,NOT_CLOSED,null}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.p.Parser: Parsed DATA frame body from java.nio.HeapByteBuffer[pos=77 lim=77 cap=8192]
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Connection: Dequeued new task null
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ChannelEndPoint: filled 0 SocketChannelEndPoint@11f9258{/10.53.133.90:16800<->/10.43.18.48:49325,OPEN,fill=-,flush=-,to=0/30000}{io=0/0,kio=0,kro=1}->HTTP2ClientConnection@2b4243
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Connection: Filled 0 bytes
2019-02-08 12:32:46,461 DEBUG o.e.j.i.FillInterest: interested FillInterest@1645d97{org.eclipse.jetty.http2.HTTP2Connection$FillableCallback@1bbd194}
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ChannelEndPoint: changeInterests p=false 0->1 for SocketChannelEndPoint@11f9258{/10.53.133.90:16800<->/10.43.18.48:49325,OPEN,fill=FI,flush=-,to=0/30000}{io=0/1,kio=0,kro=1}->HTTP2ClientConnection@2b4243
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: Queued change org.eclipse.jetty.io.ChannelEndPoint$1@1811885 on org.eclipse.jetty.io.ManagedSelector@1fcc8aa id=0 keys=1 selected=0 updates=0
2019-02-08 12:32:46,461 DEBUG o.e.j.u.t.ReservedThreadExecutor: ReservedThreadExecutor@12ecab3{s=2/2,p=0}@63440d waiting
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: Selector sun.nio.ch.WindowsSelectorImpl@2ff3f8 woken up from select, 0/1 selected
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: Selector sun.nio.ch.WindowsSelectorImpl@2ff3f8 processing 0 keys, 1 updates
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: updateable 1
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: update org.eclipse.jetty.io.ChannelEndPoint$1@1811885
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ChannelEndPoint: Key interests updated 0 -> 1 on SocketChannelEndPoint@11f9258{/10.53.133.90:16800<->/10.43.18.48:49325,OPEN,fill=FI,flush=-,to=0/30000}{io=1/1,kio=1,kro=1}->HTTP2ClientConnection@2b4243
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: updates 0
2019-02-08 12:32:46,461 DEBUG o.e.j.i.ManagedSelector: Selector sun.nio.ch.WindowsSelectorImpl@2ff3f8 waiting on select
2019-02-08 12:32:46,492 DEBUG o.a.j.v.ViewResultsFullVisualizer: Expanded: [[Root]]
2019-02-08 12:32:46,492 DEBUG o.a.j.v.ViewResultsFullVisualizer: Elements: [Root]
2019-02-08 12:32:46,507 ERROR c.b.j.h.s.HTTP2SampleResult: Error while await for response
java.util.concurrent.TimeoutException: null
	at java.util.concurrent.CompletableFuture.timedGet(Unknown Source) ~[?:1.8.0_161]
	at java.util.concurrent.CompletableFuture.get(Unknown Source) ~[?:1.8.0_161]
	at com.blazemeter.jmeter.http2.sampler.HTTP2Connection.awaitResponses(HTTP2Connection.java:197) ~[jmeter-bzm-http2-1.4.jar:?]
	at com.blazemeter.jmeter.http2.sampler.HTTP2Request.lambda$waitAllResponses$0(HTTP2Request.java:563) ~[jmeter-bzm-http2-1.4.jar:?]
	at java.util.HashMap$Values.forEach(Unknown Source) [?:1.8.0_161]
	at com.blazemeter.jmeter.http2.sampler.HTTP2Request.waitAllResponses(HTTP2Request.java:561) [jmeter-bzm-http2-1.4.jar:?]
	at com.blazemeter.jmeter.http2.sampler.HTTP2Request.threadFinished(HTTP2Request.java:524) [jmeter-bzm-http2-1.4.jar:?]
	at org.apache.jmeter.threads.JMeterThread$ThreadListenerTraverser.addNode(JMeterThread.java:682) [ApacheJMeter_core.jar:4.0 r1823414]
	at org.apache.jorphan.collections.HashTree.traverseInto(HashTree.java:994) [jorphan.jar:4.0 r1823414]
	at org.apache.jorphan.collections.HashTree.traverseInto(HashTree.java:995) [jorphan.jar:4.0 r1823414]
	at org.apache.jorphan.collections.HashTree.traverse(HashTree.java:977) [jorphan.jar:4.0 r1823414]
	at org.apache.jmeter.threads.JMeterThread.threadFinished(JMeterThread.java:649) [ApacheJMeter_core.jar:4.0 r1823414]
	at org.apache.jmeter.threads.JMeterThread.run(JMeterThread.java:305) [ApacheJMeter_core.jar:4.0 r1823414]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_161]
2019-02-08 12:32:46,507 DEBUG o.e.j.u.c.AbstractLifeCycle: stopping org.eclipse.jetty.http2.client.HTTP2Client@16e567b
2019-02-08 12:32:46,507 DEBUG o.e.j.u.c.AbstractLifeCycle: stopping HTTP2ClientSession@1e7f928{l:/10.43.18.48:49325 <-> r:/10.53.133.90:16800,sendWindow=65535,recvWindow=16776280,streams=2,NOT_CLOSED,null}
2019-02-08 12:32:46,507 DEBUG o.e.j.h.HTTP2Session: Sending GoAwayFrame@9c5d7d,0/NO_ERROR/stop/LOCALLY_CLOSED
2019-02-08 12:32:46,507 DEBUG o.e.j.h.HTTP2Flusher: Appended GoAwayFrame@9c5d7d,0/NO_ERROR/stop/LOCALLY_CLOSED, frames=1
2019-02-08 12:32:46,507 DEBUG o.e.j.h.HTTP2Flusher: Flushing HTTP2ClientSession@1e7f928{l:/10.43.18.48:49325 <-> r:/10.53.133.90:16800,sendWindow=65535,recvWindow=16776280,streams=2,LOCALLY_CLOSED,GoAwayFrame@9c5d7d,0/NO_ERROR/stop/LOCALLY_CLOSED}
2019-02-08 12:32:46,507 DEBUG o.e.j.h.HTTP2Flusher: Processing GoAwayFrame@9c5d7d,0/NO_ERROR/stop/LOCALLY_CLOSED
2019-02-08 12:32:46,507 DEBUG o.e.j.h.HTTP2Session: Generated GoAwayFrame@9c5d7d,0/NO_ERROR/stop/LOCALLY_CLOSED
2019-02-08 12:32:46,507 DEBUG o.e.j.h.HTTP2Flusher: Writing 1 buffers (21 bytes) for 1 frames [GoAwayFrame@9c5d7d,0/NO_ERROR/stop/LOCALLY_CLOSED]
2019-02-08 12:32:46,507 DEBUG o.e.j.i.ChannelEndPoint: flushed 21 SocketChannelEndPoint@11f9258{/10.53.133.90:16800<->/10.43.18.48:49325,OPEN,fill=FI,flush=W,to=46/30000}{io=1/1,kio=1,kro=1}->HTTP2ClientConnection@2b4243
2019-02-08 12:32:46,507 DEBUG o.e.j.i.WriteFlusher: Flushed=true written=21 remaining=0 WriteFlusher@1fdd6e8{WRITING}->null
2019-02-08 12:32:46,507 DEBUG o.e.j.h.HTTP2Flusher: Written 1 frames for [GoAwayFrame@9c5d7d,0/NO_ERROR/stop/LOCALLY_CLOSED]
2019-02-08 12:32:46,507 DEBUG o.e.j.i.AbstractEndPoint: shutdownOutput SocketChannelEndPoint@11f9258{/10.53.133.90:16800<->/10.43.18.48:49325,OPEN,fill=FI,flush=-,to=0/30000}{io=1/1,kio=1,kro=1}->HTTP2ClientConnection@2b4243
2019-02-08 12:32:46,507 DEBUG o.e.j.h.HTTP2Session: Disconnecting HTTP2ClientSession@1e7f928{l:/10.43.18.48:49325 <-> r:/10.53.133.90:16800,sendWindow=65535,recvWindow=16776280,streams=2,LOCALLY_CLOSED,GoAwayFrame@9c5d7d,0/NO_ERROR/stop/LOCALLY_CLOSED}

**Comment by ranjithpp** [comment]
![image](https://user-images.githubusercontent.com/47220450/52463926-dcc16600-2b9e-11e9-92d9-b66e5412041e.png)
trace also shows response received, not able to track where the problem is

**Comment by sbordet** [comment]
The logs show:

```
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Session: Received DataFrame@9616e0#35{length:52,end=true}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.FlowControlStrategy: Data received, 52 bytes, updated session recv window 16776332 -> 16776280 for HTTP2ClientSession@1e7f928{l:/10.43.18.48:49325 <-> r:/10.53.133.90:16800,sendWindow=65535,recvWindow=16776280,streams=3,NOT_CLOSED,null}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.FlowControlStrategy: Data received, 52 bytes, updated stream recv window 65535 -> 65483 for HTTP2Stream@ca613#35{sendWindow=65535,recvWindow=65483,reset=false,LOCALLY_CLOSED,age=101,attachment=null}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Stream: Update close for HTTP2Stream@ca613#35{sendWindow=65535,recvWindow=65483,reset=false,LOCALLY_CLOSED,age=102,attachment=null} close=true local=false
2019-02-08 12:32:46,461 DEBUG o.e.j.h.HTTP2Session: Removed local HTTP2Stream@ca613#35{sendWindow=65535,recvWindow=65483,reset=false,CLOSED,age=102,attachment=null}
2019-02-08 12:32:46,461 DEBUG o.e.j.h.FlowControlStrategy: Data consumed, 52 bytes, session recv window level 936/8388608 for HTTP2ClientSession@1e7f928{l:/10.43.18.48:49325 <-> r:/10.53.133.90:16800,sendWindow=65535,recvWindow=16776280,streams=2,NOT_CLOSED,null}
```

In particular:

* the `DataFrame` has `end=true`, so it's the last data frame.
* the data frame has been passed to the application which has succeeded the callback (`Data consumed, 52 bytes...`).

From the Jetty point of view all is normal.

I'm guessing it's a bug in JMeter, or perhaps a wrong configuration (timeouts too short).

**Comment by ranjithpp** [comment]
That was a super fast response! Thanks.
I have configured multiple timeout values and even set to NULL, But result is same. if it is a jmeter/http2 plugin issue, how do i raise it ?

**Comment by sbordet** [comment]
You need to raise an issue to the JMeter project: https://jmeter.apache.org/issues.html

**Comment by joakime** [comment]
or the jmeter http/2 plugin site - https://github.com/Blazemeter/jmeter-http2-plugin

**Comment by pmouawad** [comment]
Hello @sbordet ,
This is not a JMeter issue.
The sampler mentioned here is a plugin and is not provided by us.
This is the right target:

https://groups.google.com/forum/#!forum/jmeter-plugins/

Regards
